text/0000-min-rust-version.md

```diff
@@ -0,0 +1,180 @@
+
+It will be an error to use `rust="1.27"` and `edition="2018"`, but `rust="1.40"`
+and `edition="2015"` is a valid combination.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+The describe functionality can be introduced in several stages:
+
+
+## First stage: dumb field
+
+At first the `rust` field can be simply a declarative optional field without any
+functionality behind it. The reason for it is to reduce implementation cost of
+the first stage to the minimum and ideally ship it as part of Rust 2018.
+It will also allow crate authors who care about MSRV to start mark their crates
+early.
+
+
+## Second stage: `cargo publish` check
+
+The next step is for `cargo publish` to require use of the toolchain specified
+in the `rust` field, for example crates with:
+- `rust="stable"` can be published only with a stable toolchain, though not
+necessarily with the latest one. Cargo will insert toolchain version before
+publishing the crate as was described in the "guide-level explanation".
+- `rust="nightly"` can be published only with a nightly toolchain. If finer
+grained "nightly: ..." (see "nightly versions" section) is selected, then one
+of the selected Nightly versions will have to be used.
+- `rust="1.30"` can be published only with (stable) Rust 1.30, even if it's
+not the latest stable Rust version.
+
+Using the usual build check `cargo publish` will verify that crate indeed can be
+built using specified MSRV. This check can be used with exisiting `--no-verify`
+option.
+
+## Third stage: versions resolution
+
+`rust` field becomes required and cargo will add it as a constraint to dependency
+versions resolution. If user uses e.g. Rust 1.40 and uses crate `foo = "0.2"`, but
+all selected versions of `foo` specify MSRV e.g. equal 1.41 or bigger (or even
+nightly) `cargo` will issue an error.
+
+`rust` field value will be checked as well, on crate build `cargo` will check if
+all upstream dependencies can be built with the specified MSRV. (i.e. it will
+check if there is exists solution for given crates and Rust versions constraints)
+
+Yanked crates will be ignored in this process.
+
+Implementing this functionality hopefully will allow to close the debate regarding
+MSRV handling in crate versions and will allow crate authors to feel less
+restrictive about bumping their crate's MSRV. (though it can be a usefull
+convention for post-1.0 crates to bump minor version on MSRV change to allow
+publishing backports which fix serious issues using patch version)
+
+## Extension: nightly versions
+
+For some bleeding-edge crates which experience frequent breaks on Nightly updates
+(e.g. `rocket`) it can be useful to specify exact Nightly version(s) on which
+crate can be built. One way to achieve this is by using the following syntax:
+- single version: rust = "nightly: 2018-01-01"
+- enumeration: "nightly: 2018-01-01, 2018-01-15"
+- (inclusive) range: "nightly: 2018-01-01..2018-01-15"
```

Comment:
Using `..` as inclusive range in Rust is gonna be confusing. Could we incorporate the [semver operators](https://doc.rust-lang.org/cargo/reference/specifying-dependencies.html#inequality-requirements) instead?

```
nightly: >=2018-01-01, <=2018-01-15
```

(every new nightly is considered a major version for sure.)

Reply:
I am not particularly attached to the proposed syntax (I think the main use will be exact versions, i.e. "nightly: 2018-01-01", other variants probably will not be used much), so probably using semver operators will be indeed a better choice.

Though, after some thought I started to worry about discrepancy between how "stable" and "nightly" work. I think it will be better to interpret "nightly" as `cargo publish` will insert version of the currently used nightly toolchain, and use "nightly: *" for all nightlies. `cargo init` should probably use the latter variant on nightly. What do you think?

Reply:
Sounds good to me, though it makes me think if it makes sense to mix both stable and nightly in the same line?

```
rust = "1.26, nightly: >=2017-01-01"
```

Reply:
Hm, for what use-cases? If code enables nightly features it can't work on stable, and if it works on stable it *should* work on nightly. (excluding potential regressions of course, which I think we can safely ignore)

Reply:
@newpavlov `num-traits` uses a build script to determine whether to enable `i128`, which covers both stable and nightly. Though the said build script could emit an error itself meaning there's probably no real use case 🙃.